compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -37,6 +38,10 @@ pub enum MemoryKind<T> {
     Stack,
     /// Memory allocated by `caller_location` intrinsic. Error if ever deallocated.
     CallerLocation,
+    /// The base allocation of a static.
+    /// When this is read from, we need to error, because otherwise we could
+    /// invent values from nothing (e.g. `static FOO: Thing = FOO;`)
+    StaticRoot(DefId),
```

Comment:
Do we absolutely need to error here? If you do nothing, this will result in an uninitialized static, which will fail the validity check for almost all statics.

I guess it's weird though that you can get a "UB" error without writing any unsafe...

Reply:
I don't know how else to avoid the `static FOO: ZstWithPrivateFields = FOO;` case being allowed outside the module of the Zst

Reply:
Ah, good point.

Reply:
But we don't need to know the DefId, do we? Is that used anywhere?

Reply:
we use it in `before_memory_read` to cause the query cycle on self-defining ZST statics